lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1363,41 +1366,88 @@ static void diagnoseImplicitSelfUseInClosure(TypeChecker &TC, const Expr *E,
+        if (isImplicitSelfParamUse(DSCE->getBase()) &&
             isa<DeclRefExpr>(DSCE->getFn())) {
           auto MethodExpr = cast<DeclRefExpr>(DSCE->getFn());
+          memberLoc = DSCE->getLoc();
           TC.diagnose(DSCE->getLoc(),
                       diag::method_call_in_closure_without_explicit_self,
-                      MethodExpr->getDecl()->getBaseName().getIdentifier())
-            .fixItInsert(DSCE->getLoc(), "self.");
-          return { false, E };
+                      MethodExpr->getDecl()->getBaseName().getIdentifier());
         }
-
+      
+      // This error can be fixed by either capturing self explicitly (if in an
+      // explicit closure), or referencing self explicitly.
+      if (memberLoc.isValid()) {
+        if (auto *CE = dyn_cast<ClosureExpr>(ACE)) {
+          // If we've already captured something with the name "self" other than
+          // the actual self param, offer special diagnostics.
+          if (auto *VD = CE->getCapturedSelfDecl()) {
+            // Either this is a weak capture of self...
+            if (VD->getType()->is<WeakStorageType>()) {
+              TC.diagnose(VD->getLoc(), diag::note_self_captured_weakly);
+            // ...or something completely different.
+            } else {
+              TC.diagnose(VD->getLoc(), diag::note_other_self_capture);
+            }
+            // Bail on the rest of the diagnostics. Offering the option to
+            // capture 'self' explicitly will result in an error, and using
+            // 'self.' explicitly will be accessing something other than the
+            // self param.
+            // FIXME: We could offer a special fixit in the [weak self] case to insert 'self?.'...
+            return { false, E };
+          }
+          
+          auto diag = TC.diagnose(CE->getLoc(),
+                                  diag::note_capture_self_explicitly);
+          // If there's a token immediately following the opening brace of the
+          // closure, we may need to pad the fix-it with a space.
+          auto nextLoc = CE->getLoc().getAdvancedLoc(1);
+          auto next = Lexer::getTokenAtLocation(TC.Context.SourceMgr,
+                                                CE->getLoc().getAdvancedLoc(1));
+          std::string trailing = next.getLoc() == nextLoc ? " " : "";
+          if (CE->getInLoc().isInvalid())
+            diag.fixItInsertAfter(CE->getLoc(), " [self] in" + trailing);
+          else if (CE->getBracketRange().isInvalid())
+            diag.fixItInsertAfter(CE->getLoc(), " [self]" + trailing);
+          else {
+            // Insert into an existing capture list
+            auto range = CE->getBracketRange();
+            if (range.Start.getAdvancedLoc(1) == range.End)
+              diag.fixItInsertAfter(range.Start, "self");
+            else
+              diag.fixItInsertAfter(range.Start, "self, ");
+          }
+        }
+        TC.diagnose(memberLoc, diag::note_reference_self_explicitly)
```

Comment:
That seems reasonable to me. Initially had the other fix-it first since the capture list occurs syntactically "before" the actual member usage, but if the ordering doesn't matter in that respect I'm not attached to the ordering used here.

Reply:
For teaching, capture list is better suited to crystallize the concept of capturing. "Explicit .self is confusing" is one of the reasons for this SR. Maybe leave this to community discussion?